### PR TITLE
Fix `redistribute_numeric_pairs` for negative numbers

### PR DESCRIPTION
**Pull request by tybug**
This was likely an issue introduced by my refactor of the shrinker for the typed choice sequence (and discovered by investigating benchmark results for the TCS paper :sweat_smile:).

### CODE REVIEW

File: hypothesis-python/tests/conjecture/test_shrinker.py [hunk 1]
```diff
@@ -659,3 +669,16 @@ def shrinker(data: ConjectureData):
...
+    # redistributing should redistribute towards shrink_towards, not 0
+    @shrinking_from([2, 18])
+    def shrinker(data: ConjectureData):
+        v1 = data.draw_integer(shrink_towards=5)
+        v2 = data.draw_integer(shrink_towards=5)
+        if v1 + v2 == 20:
+            data.mark_interesting(interesting_origin())
+
+    shrinker.fixate_shrink_passes([ShrinkPass(shrinker.redistribute_numeric_pairs)])
+    assert shrinker.choices == (5, 15)
```
Comment (Zac-HD) [inline hypothesis-python/tests/conjecture/test_shrinker.py]: I'd like to generalize this test to include variation across int/float, towards negative/zero/positive, towards below/within/above a constrained range, and a pair below/spanning/above the reference point.

There are just enough edge cases here that I'll feel a lot better if we cover _all_ the bases in our test.

Comment (Liam-DeVoe) [reply hypothesis-python/tests/conjecture/test_shrinker.py]: writing a general float test for this is hard since the minimal input is no longer `(0, shrink_towards)` with our custom float ordering. An explicit test for floats should be good enough.

**Review by Zac-HD — APPROVED** [review]
We still don't have explicit tests for `redistribute_numeric_pairs` with one int and one float, but otherwise looks great and I'm happy for you to merge with or without such a test.
